kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterContext.java

```diff
@@ -94,8 +95,42 @@ public interface FilterContext {
      * @see io.kroxylicious.proxy.filter Thread Safety
      */
     @NonNull
+    @Deprecated(forRemoval = true, since = "0.3.0")
     <T extends ApiMessage> CompletionStage<T> sendRequest(short apiVersion, @NonNull ApiMessage request);
 
+    /**
+     * Send a request from a filter towards the broker.   The response to the request will be
+     * made available to the filter asynchronously, by way of the {@link CompletionStage}.  The
+     * CompletionStage will contain the response header and response body, or null of the
+     * request does not have a response.
+     * <h4>Header</h4>
+     * <p>Kroxylicious will automatically create a suitable header for the request, defaulting the API version to
+     * the highest request API version possible, respecting the result of the last ApiVersion negotiation that
+     * occurred on the connection.</p>
+     * <p>To customize the header (for instance, to set a {@link RequestHeaderData#clientId()}, pass a consumer and
+     * use the action to modify the fields of the header (for instance {@link RequestHeaderData#setClientId(String)}.</p>
+     * <h4>Filtering</h4>
+     * <p> The request will pass through all filters upstream of the filter that invoked the operation,
+     * invoking them, but not itself.  Similarly, the response will pass through all filters upstream
+     * of the filter that invoked the operation, invoking them, but not itself. The response does not
+     * pass through filters downstream.
+     * </p>
+     * <h4>Chained Computation stages</h4>
+     * <p>Default and asynchronous default computation stages chained to the returned
+     * {@link java.util.concurrent.CompletionStage} are guaranteed to be executed by the thread associated with the
+     * connection. See {@link io.kroxylicious.proxy.filter} for more details.
+     * </p>
+     *
+     * @param <M>           The type of the response
+     * @param headerMutator The consumer used to mutate the header, may be null.
+     * @param request       The request to send.
+     * @return CompletionStage that will yield the response.
+     * @see io.kroxylicious.proxy.filter Thread Safety
+     */
+    @NonNull
+    <M extends ApiMessage> CompletionStage<ResponseHeaderAndApiMessage<M>> sendRequest(@Nullable Consumer<RequestHeaderData> headerMutator,
```

Comment:
I needed to switch from the caller passing a `header` to the caller passing a `headerMutator`.  The behaviour I want is the Kroxylicious to defect to the highest support version, unless the user overrides.  If the argument is a RHD, we can't achieve that, as we can't distinguish between an uninitialised version and a desire for version 0.  The field is a `short`.

By having the caller pass a mutating consumer, they can apply the changes to the object. What do you think to this approach?

Reply:
Hmmm. It feels quite an awkward API. I wonder if it would make more sense for the runtime to provide `Context. createHeader()` which defaulted the version and allowed filter authors to replace it and then sendRequest just forwards what it gets?

Reply:
> The behaviour I want is the Kroxylicious to default to the highest support version

This sounds a bit risky. It is true that we could pick a version that both Kroxylicious and the Proxied Cluster support but since we are likely working with a message constructed by the Filter the author has to take care to use fields that are compatible with the api version. 

So you could have awkward effects like everything working in integration tests against the latest Kafka version, but it fails when we proxy an older cluster because the Author has used fields that are only available in the latest version.

Reply:
I like the `Context.createHeader(apiKey)` idea.  That appears much nice.  So I think the natural consequence of this is 
 that in sendRequest(header, request) the header becomes required.     We can have Context.createHeader(apiKey) use the Guarding idea to protect the object from changes such as an inconsistent API key.

Reply:
> So you could have awkward effects like everything working in integration tests against the latest Kafka version

You are right.  The idea of defaulting the highest support version is duff.  It would actually run contrary these words in the [protocol guide](https://kafka.apache.org/36/protocol.html#:~:text=The%20intended%20upgrade%20path%20is,gradually%20be%20taken%20advantage%20of.) as there is the possibility that Kroxylicious filters would fail as the broker update was rolled out.  That would create a big bang upgrade issue.

> The intended upgrade path is that new features would first be rolled out on the server (with the older clients not making use of them) and then as newer clients are deployed these new features would gradually be taken advantage of.

so I think that means the API for creating a header should be:

```java
Context.createHeader(apiKey, version)
```

Reply:
I tried the `Context.createHeader(apiKey, version)` idea.  In the end, I didn't like it.   I didn't think the benefits outweighed the disadvantage of the increased API surface.

The caller can use:
```
   context.sendRequest(new RequestHeaderData().setRequestApiVersion(x));
```

I've made `#sendRequest` unconditionally overwrite the requestRequestApiKey.

I also decided to drop the clientId default `filtername + objectid`.  I don't think the defaulting is too useful and it won't play nicely with Kafka Quotas (there's no pattern matching).